In [1]:
import os
import numpy as np
import cv2

## Exploration in CelebA-HQ dataset and Selecting one female and one male

In [2]:
input_path = "celebA-HQ/"


identity_map = {}
file_names = os.listdir(input_path)

with open("identity_CelebA.txt") as f:
    for line in f.readlines():
        img_name, identity = line.split()
        if(img_name not in file_names):
            continue
        if identity not in identity_map:
            identity_map[identity] = [img_name]
        else:
            identity_map[identity].append(img_name)

# for img in identity_map["1"]:
#     img_path = os.path.join(input_path, img)
#     img = cv2.imread(img_path)
#     cv2.imshow('test', img)
#     cv2.waitKey(0)




# for img in identity_map[female]:
#     img_path = os.path.join(input_path, img)
#     output_path = os.path.join(male_output, img)
#     img = cv2.imread(img_path)
#     cv2.imshow('test', img)
#     cv2.waitKey(0)

# for img in identity_map[female]:
#     img_path = os.path.join(input_path, img)
#     output_path = os.path.join(female_output, img)
#     img = cv2.imread(img_path)
#     cv2.imwrite(output_path, cv2.blur(img, (10,10) ))

In [3]:
identity = list(identity_map.keys())
print(len(identity))
print(sorted(identity_map, key=lambda key: len(identity_map[key]), reverse=True))
print(len(identity_map['4532']))

6217
['861', '5247', '303', '4027', '5884', '731', '4532', '3172', '4437', '248', '81', '259', '800', '1852', '318', '7137', '6556', '5198', '4154', '394', '4914', '5018', '1135', '121', '8511', '2858', '9315', '239', '1657', '4537', '4696', '6676', '3705', '6461', '3031', '1422', '1672', '674', '3817', '1499', '6574', '9245', '2603', '2955', '4388', '7002', '5158', '5805', '6440', '1362', '4268', '6624', '3648', '5611', '8702', '7331', '1944', '2909', '94', '6858', '5318', '5221', '5222', '2921', '7286', '313', '447', '818', '6833', '1977', '1927', '3347', '3092', '6708', '6958', '1145', '4811', '5795', '1281', '3421', '7168', '138', '273', '160', '2879', '5024', '931', '7210', '9610', '9570', '9700', '174', '3788', '6147', '8962', '6576', '5142', '2792', '6063', '3993', '1590', '2002', '7360', '8776', '1137', '3839', '3835', '3802', '7022', '3870', '5344', '1128', '9542', '1938', '1359', '6688', '5601', '9211', '5157', '7475', '1962', '1629', '230', '7613', '6532', '320', '353', '89'

The number of identities in CelebA-HQ is 6217. The largest number of female identity is 28, whose ID is '861'; The largest number of male identity is 24, whose ID is '4532'

In [6]:
female = "861"
male = "4532"
female_output = "./female_origin"
male_output = "./male_origin"
for img in identity_map[male]:
    img_path = os.path.join(input_path, img)
    out_path = os.path.join(male_output, img)
    cur = cv2.imread(img_path)
    cv2.imwrite(out_path, cur)
for img in identity_map[female]:
    img_path = os.path.join(input_path, img)
    out_path = os.path.join(female_output, img)
    cur = cv2.imread(img_path)
    cv2.imwrite(out_path, cur)

## Test Gaussian Blur


In [2]:
import os
import numpy as np
import cv2
from torchvision import transforms
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader, random_split
from torchvision.utils import save_image

In [15]:
female_origin = "./female_origin"
male_origin = "./male_origin"
female_blur = "./female_blur"
male_blur = "./male_blur"

level_1 = 15
level_2 = 25
level_3 = 45

# FeMale
for img_name in os.listdir(female_origin):
    img_path = os.path.join(female_origin, img_name)
    img = cv2.imread(img_path)

    blurred_1 = cv2.GaussianBlur(img, (level_1, level_1), 0)
    output_path = os.path.join(female_blur, "level_1/", img_name)
    cv2.imwrite(output_path, blurred_1)

    blurred_2 = cv2.GaussianBlur(img, (level_2, level_2), 0)
    output_path = os.path.join(female_blur, "level_2/", img_name)
    cv2.imwrite(output_path, blurred_2)

    blurred_3 = cv2.GaussianBlur(img, (level_3, level_3), 0)
    output_path = os.path.join(female_blur, "level_3/", img_name)
    cv2.imwrite(output_path, blurred_3)

# Male
for img_name in os.listdir(male_origin):
    img_path = os.path.join(male_origin, img_name)
    img = cv2.imread(img_path)

    blurred_1 = cv2.GaussianBlur(img, (level_1, level_1), 0)
    output_path = os.path.join(male_blur, "level_1/", img_name)
    cv2.imwrite(output_path, blurred_1)

    blurred_2 = cv2.GaussianBlur(img, (level_2, level_2), 0)
    output_path = os.path.join(male_blur, "level_2/", img_name)
    cv2.imwrite(output_path, blurred_2)

    blurred_3 = cv2.GaussianBlur(img, (level_3, level_3), 0)
    output_path = os.path.join(male_blur, "level_3/", img_name)
    cv2.imwrite(output_path, blurred_3)

## Split into Train, Test, and Validation

In [16]:
clear_transforms = transforms.Compose([
    # transforms.RandomResizedCrop(224),
    # transforms.RandomHorizontalFlip(0.1),
    # transforms.ColorJitter(brightness=0.2, contrast=0.2, saturation=0.2, hue=0.1),
    # transforms.RandomRotation(20),
    # transforms.RandomVerticalFlip(0.1),
    transforms.ToTensor(),
    # transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])  # Assuming ImageNet mean and std
])

dataset = ImageFolder(root='male_origin', transform=clear_transforms)

train_size = int(0.7 * len(dataset))  # 70% for training
val_size = int(0.15 * len(dataset))   # 15% for validation
test_size = len(dataset) - train_size - val_size  # Remaining for testing


train_size = int(0.7 * len(dataset))  # 70% for training
val_size = int(0.15 * len(dataset))   # 15% for validation
test_size = len(dataset) - train_size - val_size  # Remaining for testing

train_dir = "datasets/clear/train"
val_dir = "datasets/clear/val"
test_dir = "datasets/clear/test"

train_dataset, val_dataset, test_dataset = random_split(dataset, [train_size, val_size, test_size])

for dataset, output_folder in [(train_dataset, train_dir), (val_dataset, val_dir), (test_dataset, test_dir)]:
    for idx in range(len(dataset)):
        image, _ = dataset[idx]
        output_path = os.path.join(output_folder, f"{idx}.jpg")
        save_image(image, output_path)

FileNotFoundError: Couldn't find any class folder in male_origin.